packages/backend-common/src/reading/tree/ArchiveResponse.ts

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright 2020 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import os from 'os';
+import tar, { Parse, ParseStream, ReadEntry } from 'tar';
+import path from 'path';
+import fs from 'fs-extra';
+import { Readable, pipeline as pipelineCb } from 'stream';
+import { promisify } from 'util';
+import concatStream from 'concat-stream';
+import { ReadTreeResponse, File, ReadTreeResponseDirOptions } from '../types';
+
+// Tar types for `Parse` is not a proper constructor, but it should be
+const TarParseStream = (Parse as unknown) as { new (): ParseStream };
+
+const pipeline = promisify(pipelineCb);
+
+/**
+ * Wraps a tar archive stream into a tree response reader.
+ */
+export class ArchiveResponse implements ReadTreeResponse {
+  private read = false;
+
+  constructor(
+    private readonly stream: Readable,
+    private readonly subPath: string,
+    private readonly workDir: string = os.tmpdir(),
+    private readonly filter?: (path: string) => boolean,
+  ) {
+    if (subPath) {
+      if (!subPath.endsWith('/')) {
+        this.subPath += '/';
+      }
+      if (subPath.startsWith('/')) {
+        throw new TypeError(
+          `ArchiveResponse subPath must not start with a /, got '${subPath}'`,
```

Comment:
Would it just be too confusing to silently cut out leading slashes?

Reply:
It's really an implementation error and not a user error so figured this could help avoid mistakes. I actually did the same with the suffix, but that ended up being a bit more input dependent so switched things around. Could definitely switch to that in the future if there's a reason tho.